src/graphnet/data/pipeline.py

```diff
@@ -0,0 +1,173 @@
+        return
+    def _load_modules(self, module_dict):
+        """Loads each module in _module_dict
+
+        Args:
+            module_dict (dict): dictionary of GNN ModuleList
+
+        Returns:
+            dict: module_dict with "loaded_module" field added.
+        """
+        for target in module_dict.keys():
+            model = torch.load(module_dict[target]['path'], map_location = 'cpu', pickle_module = dill)
+            model._gnn.to(self._device)
+            model._device = self._device
+            model.to(self._device)
+            model.eval()
+            module_dict[target]['loaded_module'] = model
+        return module_dict
+
+    def _make_dataloader(self, database, pulsemap):
+        dataloader = make_dataloader(db = database, pulsemaps = pulsemap, features = self._features, truth = self._truth, batch_size = self._batch_size, shuffle = False, selection = None, num_workers = self.n_workers, persistent_workers= False)
+        return dataloader
+
+    def _combine_outputs(self, dataframes):
+        return reduce(lambda x, y: pd.merge(x, y, on = 'event_no'), dataframes)
+
+    def _get_outdir(self, database):
+        if self._outdir == None:
+            database_name = database.split('/')[-3]
+            outdir = database.split(database_name)[0] + database_name +  '/pipelines/' + self._pipeline_name
+        else:
+            outdir = self._outdir
+        return outdir
+
+    def _get_truth(self,database):
+        with sqlite3.connect(database) as con:
+            query = 'SELECT * FROM truth'
+            truth = pd.read_sql(query,con)
+        return truth
+
+    def _get_retro(self,database):
+        try:
+            with sqlite3.connect(database) as con:
+                query = 'SELECT * FROM %s'%self._retro_table_name
+                retro = pd.read_sql(query,con)
+            return retro
+        except:
+            print('%s table does not exist'%self._retro_table_name)
+            return
+
+
+    def _make_pipeline_database(self,outdir, df, original_database):
+        os.makedirs(outdir)
+        pipeline_database = outdir + '/%s.db'%self._pipeline_name
+        truth = self._get_truth(original_database)
+        retro = self._get_retro(original_database)
+
+        self._save_to_sql(df, 'reconstruction', pipeline_database)
+        self._save_to_sql(truth, 'truth', pipeline_database)
+        if isinstance(retro, pd.DataFrame):
+            self._save_to_sql(retro, self._retro_table_name, pipeline_database)    
+        return
+        
+
+    def _save_to_sql(self, df, table_name, pipeline_database):
+        print('SUBMITTING %s'%table_name)
+        self._create_table(pipeline_database,  table_name, df)
+        engine = sqlalchemy.create_engine('sqlite:///' + pipeline_database)
+        df.to_sql(table_name, con=engine, index=False, if_exists='append')
+        engine.dispose()
+        print('%s SUBMITTED'%table_name)
+        return        
```

Comment:
This could probably also be generalised into a common utility method (see comment below).

Reply:
moved